### ISSUE TITLE
[rt.openssl.org #4685] [PATCH] Add missing prototype for FIPS callback

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#2533]
Migrated from [rt.openssl.org#4685](https://rt.openssl.org/Ticket/Display.html?id=4685) (status was 'new')

Requestors:
* Matthias.St.Pierre@ncp-e.com

From matthias.st.pierre@ncp-e.com on 2016-09-26 13:41:45:
```
The call to FIPS_crypto_set_id_callback() was added in revision a43cfd7bb1fc681d563e,
but there is no prototype for it in <openssl/fips.h>.
---

This leads to warnings on some platforms (e.g. x86_64-ncp-linux-gnu-gcc):
o_init.c:77:5: warning: implicit declaration of function 'FIPS_crypto_set_id_callback' [-Wimplicit-function-declaration]

and to an error on iOS (clang -arch arm64):
o_init.c:77:5: error: implicit declaration of function 'FIPS_crypto_set_id_callback' is invalid in C99 [-Werror,-Wimplicit-function-declaration]


 crypto/o_init.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/crypto/o_init.c b/crypto/o_init.c
index 185841e..a399318 100644
--- a/crypto/o_init.c
+++ b/crypto/o_init.c
@@ -74,6 +74,8 @@ void OPENSSL_init(void)
 #ifdef OPENSSL_FIPS
     FIPS_set_locking_callbacks(CRYPTO_lock, CRYPTO_add_lock);
 # ifndef OPENSSL_NO_DEPRECATED
+    /* the prototype is missing in <openssl/fips.h> */
+    void FIPS_crypto_set_id_callback(unsigned long (*func)(void));
     FIPS_crypto_set_id_callback(CRYPTO_thread_id);
 # endif
     FIPS_set_error_callbacks(ERR_put_error, ERR_add_error_vdata);
-- 
2.7.3

```

From matthias.st.pierre@ncp-e.com on 2016-09-26 16:23:45:
```
The call to FIPS_crypto_set_id_callback() was added in revision a43cfd7bb1fc681d563e,
but there is no prototype for it in <openssl/fips.h>.
---

Moved the function prototype upwards, because declarations can only be placed
at the top of a function in C.


 crypto/o_init.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/crypto/o_init.c b/crypto/o_init.c
index 185841e..18bb858 100644
--- a/crypto/o_init.c
+++ b/crypto/o_init.c
@@ -58,6 +58,11 @@
 #ifdef OPENSSL_FIPS
 # include <openssl/fips.h>
 # include <openssl/rand.h>
+
+# ifndef OPENSSL_NO_DEPRECATED
+/* the prototype is missing in <openssl/fips.h> */
+void FIPS_crypto_set_id_callback(unsigned long (*func)(void));
+# endif
 #endif
 
 /*
-- 
2.7.3

```

**Comment by richsalz** [comment]
ping @snhenson  -- can we fix this without re-validating?

**Comment by mspncp** [comment]
This change does not affect the fips object module, only the fips-enabled libcrypto. So no problem with FIPS ;-)

**Comment by mattcaswell** [comment]
@mspncp Could you recreate this patch as a github PR?

**Comment by mspncp** [comment]
Already in progress ;-)

**Comment by mspncp** [comment]
See #4870

**Comment by Jan-E** [comment]
FWIW: https://github.com/openssl/openssl/pull/4870 solved my issue in https://github.com/openssl/openssl/issues/4864

**Comment by mspncp** [comment]
@richsalz this issue can be closed now. Fixed by #4870.
